### DISCUSSION TITLE
WindTerm 2.7.0 Prerelease 1

### DISCUSSION THREAD

**Discussion by kingToolbox** [kingToolbox/WindTerm#2259] [General]
# Spotlight

## Agent forwarding

![Agent](https://github.com/kingToolbox/WindTerm/assets/51009775/aab296d6-46b0-479d-a1ed-17058151236e)

#### `Agent Forwarding` is a powerful feature that enhances remote access security by allowing the use of private SSH keys locally without exposing them to remote servers. 

# Performance

#### The redesigned `Log` module, combined with further code optimization, allows `WindTerm` to run several times faster than other terminals, regardless of whether logging is enabled or not.

![benchmark](https://github.com/kingToolbox/WindTerm/assets/51009775/1ecf143d-57b4-4944-ad1f-da61be814336)
![benchmark](https://github.com/kingToolbox/WindTerm/assets/51009775/6e6b6fe3-fcb5-4088-aa66-173ea323ecba)
![benchmark](https://github.com/kingToolbox/WindTerm/assets/51009775/03334a00-ed42-4456-bf4a-4e01bc692dda)

#### The redesigned `Sftp Transfer` module enables the fastest possible transfer speeds for both large files and a large number of small files. In some cases, the speed can even be up to ten times faster than professional tools.

![benchmark](https://github.com/kingToolbox/WindTerm/assets/51009775/69ec51a2-848a-4256-a2fd-6a8f4ad2ea83)
![benchmark](https://github.com/kingToolbox/WindTerm/assets/51009775/aab10791-69df-44d3-a222-83ee3926293e)
![benchmark](https://github.com/kingToolbox/WindTerm/assets/51009775/c124f7db-e328-4208-a1c8-495e92750428)

# Usage:

### Windows

|   Version | Architecture |  Operating Systems |
| --------- | -------------- | --------------------- |
| WindTerm_2.7.0_Prerelease_1_Windows_Portable_x86_64 | x86_64 | Windows 7, Windows 8, Windows 10, Windows 11 |

#### Versions before Version 2.4.0

1. Start `WindTerm` and select the `profiles directory` and **quit**.
2. Replace the new `.wind/profiles` folder with the old `profiles` folder.
3. Restart `WindTerm` and all the sessions and configuration will be restored.

#### Versions after Version 2.4.0

1. Start `WindTerm` and select the `profiles directory` containing the old `.wind` folder.
2. It is recommended to choose the `Home Directory` to facilitate subsequent upgrades.

If the `.wind` folder needs to be moved to another system:
1. Quit `WindTerm`.
4. Replace the new `.wind` folder with the old `.wind` folder.
5. Restart `WindTerm` and all the sessions and configuration will be restored.

The **.wind** folder is cross-platform. 

# Known Issues:
- [ZMODEM] **Only support file names contains ASCII characters in Windows, otherwise the file name of the transfered file will be garbled after transmission, but it does not affect the correctness of the file content.** Linux and MacOS version does not have this problem.

# Changes in Version 2.7.0 Prerelease 1 (2024-4-2):

# Features:
- [GUI] **Supports GPU-accelerated rendering for text.**
- [GUI] **Uses full hinting for text rendering to make the rendered text more visually appealing.**
- [GUI] Automatically save `text marks` when exiting, and restore them upon next start-up.
- [GUI] Disables `rainbow brackets` on the `alt screen` to avoid conflicts with features in editors like Vim. #2111 
- [FILER] Supports displaying files within the `WSL` subsystem.
- [FILER] Supports synchronizing the current folder of the `WSL` subsystem.
- [SESSION] **Supports deleting history items from the auto-complete list using `Shift+Del`.**
- [SFTP] In the gaps between sending and receiving packets, the SFTP thread will sleep to reduce CPU usage.
- [SSH] **Perform SSH operations, including connection establishment, authentication and command execution, in an asynchronous manner.**
- [SSH] **Supports `ssh-agent` for authentication on `Windows`.**
- [SSH] **Supports `agent forwarding`, with agents including `pageant`, `ssh-agent`, `$SSH_AUTH_SOCK`, `Enviroment variable`, `Unix domain socket` etc.**
- [SSH] Strictly ensure that the opening sequence is `Shell`, `SFtp`, `Sysmon`.
- [SSH] Supports sending `signals`. And automatically sends `SIGINT` signal when `Ctrl+C` is pressed.
- [SSH] **Removes the internal proxy socket, allowing `libssh` to directly connect to the server and send/receive packets. This not only significantly improves packet transmission performance but also reduces the receive buffer by `256KiB` per session.**
- [SYSMON] Supports displaying system information of the `WSL` subsystem, including CPU, memory, time, etc.

# Improvements:
- [GUI] Significantly improves the clarity of screenshots in tab tooltips.
- [SESSION] **Significantly improved text parsing, modification, and rendering, especially to avoid extremely slow screen rendering when the text exceeds the maximum number of srollback lines.**

# Bugs:
- [APP] The application sometimes crashes when rapidly reconnecting to the server repeatedly.
- [SESSION] **When a command generates a large amount of output, pressing `Ctrl+C` does not immediately stop the command execution.**
- [SESSION] **During file transfers, keyboard input becomes sluggish or even unresponsive.**
- [SESSION] When text exceeds the maximum number of scrollback lines, the text at the bottom is sometimes incorrectly colored. 
- [WSL] In Vim on `WSL2`, uppercase letters cannot be entered using the Shift key. 

<hr /><em>This discussion was created from the release <a href='https://github.com/kingToolbox/WindTerm/releases/tag/2.7-prerelease'>WindTerm 2.7.0 Prerelease 1</a>.</em>

**Comment by JohnnyJosda** [comment]
amazing

**Comment by robot-dot-win** [comment]
Great!

**Comment by DesistDaydream** [comment]
love you

**Comment by spinachomes** [comment]
can you provide the macos installation package?

**Reply by kingToolbox** [reply]
Currently, it's only the pre-release stage. To reduce the workload, only the `Windows` version will be released to collect bugs and issues. The official release will include the `macOS` version, so please be patient. Thank you.

**Reply by stuacct071** [reply]
When will the macos version be released?

**Comment by robot-dot-win** [comment]
360 Security Software reports that there's "HEUR/QVM202.0.69DE.Malware.Gen" in "vendors\vcxsrv\xrdb.exe".

**Reply by kingToolbox** [reply]
Thank you. `vcxsrv` comes from [sourceforge: vcxsrv](https://sourceforge.net/projects/vcxsrv/files/vcxsrv/), and it has been used in the release package for a long time. You can use other antivirus software to double-check, it should be fine, because I have never received similar reports from other users.

**Reply by robot-dot-win** [reply]
It's somewhat weird - Just now I scanned it again with 360 Security, it doesn't report malware any more!

**Reply by kingToolbox** [reply]
Yes, it should be a false positive from the antivirus software. Thanks for the good news.

**Comment by guowenju** [comment]
very good

**Comment by guowenju** [comment]
I am using win11 system, but when I open the software, it does not respond. Normal when started with administrator rights

**Reply by kingToolbox** [reply]
Sorry, I don't have a `Windows 11` for testing at the moment, so I can't determine what happened. What I can think of is perhaps it doesn't have permission to use the GPU? I will investigate this issue and fix it in the next version. Please stay tuned, thank you.

**Reply by magixyu** [reply]
It works well on my Win11. 
@sfwwslm  you need to change the permission of installed directory. Then it will work.

**Reply by guowenju** [reply]
@magixyu 

Thank you, my system has been upgraded to Windows 11 24H2, v2.7.0 can work normally

**Comment by kingToolbox** [comment]
Thank you to each and every one of you. Without your support, it would be difficult to have the motivation to continuously develop new versions. Since the release of version `2.6.0`, I have been traveling and working out of town, so I have been online rarely, resulting in many accumulated issues that I haven't been able to reply to in a timely manner. I am really very sorry about that. Fortunately, most issues have actually been fixed or implemented, and I still promise to reply to every issue, so please be patient.

**Reply by darkhandz** [reply]
Hey buddy, thank you for sacrificing your time to continue updating WindTerm. It remains my favorite serial port logging tool (I rarely use SSH). Wishing you a pleasant journey!

**Reply by kingToolbox** [reply]
Thank you. Although `serial` hasn't had new features added for a long time, it hasn't been forgotten. The features currently planned for development include: real-time waveform plotting, automatically displaying received data frame by frame,  supporting parsing and colorized display of data according to protocol templates, and so on. I hope you will like them.

**Reply by darkhandz** [reply]
For me, the most desired improvement right now would be fixing the timestamp issue with serial port reception in Windterm. In recent versions, when a serial port receives an empty line, the timestamp displayed reverts to a previous time, which is a bit inconvenient. Of course, proceed according to your development schedule, but please don't forget about this issue. :)

**Comment by roke-julian-lockhart** [comment]
> https://github.com/kingToolbox/WindTerm/discussions/2259#discussion-6453190

@kingToolbox, per https://github.com/kingToolbox/WindTerm/releases/tag/2.7-prerelease#:~:text=the%20Shift%20key.-,assets,-3, do you intend to provide a Linux release?

**Reply by kingToolbox** [reply]
Fortunately, this bug only exists in `WSL 2`, and other systems are not affected. So currently there are no plans to release a `Linux` version. However, the official `2.7.0` will release a Linux version, and it shouldn't take too long.

**Reply by roke-julian-lockhart** [reply]
> https://github.com/kingToolbox/WindTerm/discussions/2259#discussioncomment-9001334

@kingToolbox, which bug do you refer to? Regardless, I don't use WSL 1 or 2 - I use a Linux OS, so I'm a little confused.

**Reply by kingToolbox** [reply]
Oh, that's fine then. I just saw that the link you quoted contained `#:~:text=the%20Shift%20key.-,assets,-3`, so I thought you were referring to issue #2111.

**Comment by xguyu** [comment]
Two-factor authentication，When SSH authenticates, the Enter Verification Code window does not pop up

**Comment by laotse-id** [comment]
👍

**Reply by fwnan** [reply]
Author, when does it support directly dragging files to transfer to the WSL system?  Just like any other Linux to use it.

**Comment by yangkzh** [comment]
![2024-04-18_103349](https://github.com/kingToolbox/WindTerm/assets/44106463/a72caa0c-4837-4bc6-996b-2d8afb679b7a)

Encountering "the remote host closed the connection" during the login process will directly cause the software to be forcibly shut down, as shown in the figure

**Comment by a19ttc29** [comment]
解决了，vim 编辑会出现乱码 。very good！

**Comment by abcxcba** [comment]
@kingToolbox 
WindTerm's functionality and performance is so good that it will probably beat other programs of its kind and become very well known in the history of the world.
I think you are a real programmer, the same kind of programmers who were great in the 1970s and 1980s, with real programming skills and thinking.
I hope that WindTerm will be better and better known and liked by more people, and I wish you all the happiness in the world.

**Comment by 1puls2is3** [comment]
@kingToolbox 

I encountered two problems while trying WindTerm.
Windows 11 Professional 23H2

2.6.1: After connecting to a virtual machine running on my computer using SSH, and executing normal commands such as cat, ls, etc. for a while, it seems that the network is getting slower and slower, and the commands are not displayed until a few seconds after they are typed in, and then it finally disconnects completely, and I can't connect to it using any other terminals either.

WindTerm 2.7.0 Prerelease 1 has a very high memory usage.
The software uses 179 megabytes of memory when it starts running.
For each additional ssh connection, the memory increases by about 100M.
But everything seems to work fine.

2.6.1 has a small memory footprint but doesn't work properly, 2.7.0_Prerelease_1 functions properly but has a very high resource footprint, even more than chrome.

**Reply by 1puls2is3** [reply]
I just tested 2.5.0 and it works fine, no problems with 2.6.1 or 2.7.0_Prerelease_1.

**Reply by 1puls2is3** [reply]
After using 2.5.0 for about 5~6 hours, I have the same problem as 2.6.1, but it seems to take longer time for the problem to appear.
The pinging of other IPs on the same network fails, and after restarting the WiFi port, the network access returns to normal.

I'm using a laptop with a wireless network.
I don't have this problem with other terminals.

**Comment by Zuolijia** [comment]
I am using win10 system with a specific build version of 17763.6054, but when I open the software, I only see a completely white interface, and it doesn't respond to anything

**Reply by LiuYanHao789** [reply]
I have the same issue. Have you solved it? However, I haven't encountered this issue on my Windows 11 system.

**Comment by VladislavGatsenko** [comment]
up

**Comment by djcry** [comment]
Hey, buddy.
After password authentication when opening the software, can password authentication be canceled when adding OneKey during normal software use? Every time I add OneKey, I have to do password authentication again, which makes me feel very distressed.
I hope to evaluate this feature.
Thank you, buddy.

**Comment by gnulux-chambe** [comment]
Hello,
very interested in this version and the support of the forwarding agent.
Do we have an idea of ​​the release schedule of the linux version please?
many thanks

**Comment by larkzhang** [comment]
I also want to join the team testing the new version. Have you considered releasing a test package for the macos version?
